src/devices/Pn5180/Pn5180.cs

```diff
@@ -0,0 +1,1496 @@
+        {
+            // Both master and slave devices must operate with the same timing.The master device
+            // always places data on the MOSI line a half cycle before the clock edge SCK, in order for
+            // the slave device to latch the data.
+            // The BUSY line is used to indicate that the system is BUSY and cannot receive any data
+            // from a host.Recommendation for the BUSY line handling by the host:
+            // 1.Assert NSS to Low
+            // 2.Perform Data Exchange
+            // 3.Wait until BUSY is high
+            // 4.Deassert NSS
+            // 5.Wait until BUSY is low
+            // Wait for the PN8150 to be ready
+            int timeout = TimeoutWaiting;
+            while (_gpioController.Read(_pinBusy) == PinValue.High)
+            {
+                if (timeout-- == 0)
+                {
+                    throw new TimeoutException($"PN8150 not ready to write");
+                }
+            }
+
+            _gpioController.Write(_pinNss, PinValue.Low);
+            Thread.Sleep(2);
+            _spiDevice.Write(toSend);
+
+            timeout = TimeoutWaiting;
+            while (_gpioController.Read(_pinBusy) == PinValue.Low)
+            {
+                if (timeout-- == 0)
+                {
+                    throw new TimeoutException($"PN8150 is still busy after writting");
+                }
+            }
+
+            _gpioController.Write(_pinNss, PinValue.High);
+        }
+
+        private void SpiRead(Span<byte> toRead)
+        {
+            // Both master and slave devices must operate with the same timing.The master device
+            // always places data on the MOSI line a half cycle before the clock edge SCK, in order for
+            // the slave device to latch the data.
+            // The BUSY line is used to indicate that the system is BUSY and cannot receive any data
+            // from a host.Recommendation for the BUSY line handling by the host:
+            // 1.Assert NSS to Low
+            // 2.Perform Data Exchange
+            // 3.Wait until BUSY is high
+            // 4.Deassert NSS
+            // 5.Wait until BUSY is low
+
+            // Wait for the PN8150 to be ready
+            int timeout = TimeoutWaiting;
+            while (_gpioController.Read(_pinBusy) == PinValue.High)
+            {
+                if (timeout-- == 0)
+                {
+                    throw new TimeoutException($"PN8150 not ready to write");
+                }
+            }
+
+            _gpioController.Write(_pinNss, PinValue.Low);
+
+            // In order to write data to or read data from the PN5180, "dummy reads" shall be
+            // performed.The Figure 8 and Figure 9 are illustrating the usage of this "dummy reads" on
+            // the SPI interface.
+            for (int i = 0; i < toRead.Length; i++)
+            {
+                _spiDevice.TransferFullDuplex(new byte[] { 0xFF }, toRead.Slice(i, 1));
```

Comment:
please preallocate this since this will allocate in a loop right now and it doesn't even change